wallet-new/test/unit/TxMetaStorageSpecs.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE RankNTypes #-}
+module TxMetaStorageSpecs (txMetaStorageSpecs) where
+
+import           Universum
+
+import           Cardano.Wallet.Kernel.DB.TxMeta
+import           Control.Exception.Safe (bracket, catchJust)
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Text.Buildable (build)
+
+import           Formatting (bprint)
+import           Serokell.Util.Text (listJsonIndent)
+import           System.Directory (getTemporaryDirectory, removeFile, withCurrentDirectory)
+import           System.IO.Error (isDoesNotExistError)
+import           Test.Hspec (shouldThrow)
+import           Test.QuickCheck (arbitrary, generate, resize, vectorOf)
+import           Util.Buildable.Hspec
+
+import qualified Cardano.Wallet.Kernel.DB.Sqlite as Storage
+
+-- | Handy combinator which yields a fresh database to work with on each spec.
+withTemporaryDb :: forall m a. (MonadIO m, MonadMask m) => (MetaDBHandle -> m a) -> m a
+withTemporaryDb action = bracket acquire release action
+    where
+       acquire :: m MetaDBHandle
+       acquire = liftIO $ do
+           tmpDir <- getTemporaryDirectory
+           withCurrentDirectory tmpDir $ do
+               -- Remove any existing db, if any.
+               catchJust (\e -> if isDoesNotExistError e then Just () else Nothing)
+                         (removeFile "txmeta_test.db")
+                         (\_ -> return ())
+               db <- openMetaDB "txmeta_test.db"
+               Storage.unsafeMigrate db
+               return db
+
+       release :: MetaDBHandle -> m ()
+       release = liftIO . closeMetaDB
+
+
+-- | Generates two 'TxMeta' which are @almost@ identical, if not in the
+-- arrangement of their inputs & outputs.
+genSimilarTxMetas :: IO (TxMeta, TxMeta)
+genSimilarTxMetas = liftIO $ do
+    inputs  <- generate (resize 5 arbitrary)
+    outputs <- generate (resize 5 arbitrary)
+    blueprint <- generate arbitrary
+    let t1 = blueprint & over txMetaInputs  (const inputs)
+                       . over txMetaOutputs (const outputs)
+    let t2 = blueprint & over txMetaInputs  (const (NonEmpty.reverse inputs))
+                       . over txMetaOutputs (const (NonEmpty.reverse outputs))
+    return (t1, t2)
+
+
+-- | Handy wrapper to be able to compare things with the 'isomorphicTo'
+-- combinator, which ignores the different order of the inputs & outputs.
+data DeepEqual = DeepEqual TxMeta
```

Comment:
if `newtype`, then you can use `GeneralizedNewtypeDeriving` to get rid of some of the boilerplate here